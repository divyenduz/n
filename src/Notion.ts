import { JSDOM } from 'jsdom'
import prettier from 'prettier'

interface NotionRuleCreate {
  element: string
  siblingOf?: string
  childOf?: string
  args: { [key: string]: string | null }
}

// type IdentifyFn = { <T>(arg: T): T }
type IdentifyFn = { (arg: string): string }

interface NotionRuleUpdate {
  element: string
  args: { [key: string]: string | null | IdentifyFn }
}

interface NotionRuleDelete {
  selectorAll: string
}

export interface NotionRule {
  type: 'create' | 'update' | 'delete'
  operation: NotionRuleCreate | NotionRuleUpdate | NotionRuleDelete
}

export class Notion {
  _rules: NotionRule[]

  constructor(rules: NotionRule[]) {
    this._rules = rules
  }

  _applyCreate(rule: NotionRule, dom: JSDOM) {
    const operation: NotionRuleCreate = rule.operation as NotionRuleCreate
    const element = dom.window.document.createElement(operation.element)
    Object.entries(operation.args).forEach(([key, value]) => {
      if (key === 'innerHTML') {
        element.innerHTML = value
      } else {
        element.setAttribute(key, value)
      }
    })
    if (operation.siblingOf) {
      const siblineElement = dom.window.document.querySelector(
        operation.siblingOf,
      )
      siblineElement.parentNode.insertBefore(
        element,
        siblineElement.nextSibling,
      )
    }
    if (operation.childOf) {
      const parentElement = dom.window.document.querySelector(operation.childOf)
      parentElement.insertBefore(element, parentElement.firstChild)
    }
    return dom.serialize()
  }

  _applyUpdate(rule: NotionRule, dom: JSDOM) {
    const operation: NotionRuleUpdate = rule.operation as NotionRuleUpdate
    const element = dom.window.document.querySelector(operation.element)
    Object.entries(operation.args).forEach(([key, value]) => {
      if (typeof value === 'string') {
        if (key === 'innerHTML') {
          element.innerHTML = value
        } else {
          element.setAttribute(key, value)
        }
      } else if (typeof value === 'function') {
        if (key === 'innerHTML') {
          element.innerHTML = value(element[key])
        } else {
          element.setAttribute(key, value(element[key]))
        }
      } else {
        throw new Error(`typeof value is unknown`)
      }
    })
    return dom.serialize()
  }

  _applyDelete(rule: NotionRule, dom: JSDOM) {
    const operation: NotionRuleDelete = rule.operation as NotionRuleDelete
    const elements = dom.window.document.querySelectorAll(operation.selectorAll)
    elements.forEach(element => {
      element.parentNode.removeChild(element)
    })
    return dom.serialize()
  }

  modifyHTML(html: string) {
    let modifyHTML = html

    this._rules.forEach(rule => {
      let dom = new JSDOM(modifyHTML)
      switch (rule.type) {
        case 'create':
          modifyHTML = this._applyCreate(rule, dom)
          break
        case 'update':
          modifyHTML = this._applyUpdate(rule, dom)
          break
        case 'delete':
          modifyHTML = this._applyDelete(rule, dom)
          break
        default:
          console.log('warn: unknown rule')
          break
      }
    })

    return prettier.format(modifyHTML, {
      parser: 'html',
    })
  }
}
