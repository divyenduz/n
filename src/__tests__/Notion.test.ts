import fs from 'fs'
import path from 'path'

import { Notion, NotionRule } from '../Notion'

describe('Test Notion', () => {
  it('should modify html based on rules', () => {
    const html = fs.readFileSync(path.join(__dirname, './fixture.html'), 'utf8')
    const rules: NotionRule[] = [
      {
        type: 'delete',
        operation: {
          selectorAll: 'style',
        },
      },
      {
        type: 'delete',
        operation: {
          selectorAll: 'link',
        },
      },
      {
        type: 'create',
        operation: {
          element: 'link',
          siblingOf: 'title',
          args: {
            href: '/style.css',
            rel: 'stylesheet',
            type: 'text/css',
          },
        },
      },
      {
        type: 'update',
        operation: {
          element: 'title',
          args: {
            innerHTML: value => {
              return `ZEBRA ` + value
            },
          },
        },
      },
    ]
    const notion = new Notion(rules)
    expect(notion.modifyHTML(html)).toMatchSnapshot()
  })
})
