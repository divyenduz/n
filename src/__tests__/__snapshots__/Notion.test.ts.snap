// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Notion should modify html based on rules 1`] = `
"<html>
  <head>
    <meta http-equiv=\\"Content-Type\\" content=\\"text/html; charset=utf-8\\" />
    <title>ZEBRA Computer Architecture</title>
    <link href=\\"/style.css\\" rel=\\"stylesheet\\" type=\\"text/css\\" />
  </head>
  <body>
    <article id=\\"bac5cb87-b09d-4fe0-9e52-27e3125aaf4a\\" class=\\"page sans\\">
      <header>
        <div class=\\"page-header-icon undefined\\">
          <span class=\\"icon\\">ðŸŒ†</span>
        </div>
        <h1 class=\\"page-title\\">Computer Architecture</h1>
        <table class=\\"properties\\">
          <tbody>
            <tr class=\\"property-row property-row-text\\">
              <th>Best&nbsp;book</th>
              <td>Computer Organization and Design</td>
            </tr>
            <tr class=\\"property-row property-row-text\\">
              <th>Best&nbsp;videos</th>
              <td>Berkeley CS&nbsp;61C</td>
            </tr>
            <tr class=\\"property-row property-row-text\\">
              <th>Why&nbsp;study?</th>
              <td>
                If you donâ€™t have a solid mental model of how a computer
                actually works, all of your higher-level abstractions will be
                brittle.
              </td>
            </tr>
          </tbody>
        </table>
      </header>
      <div class=\\"page-body\\">
        <p id=\\"301a3235-28a1-4464-9b45-9fc237cc5829\\" class=\\"\\">Books:</p>
        <p id=\\"84925743-6104-4f6c-9227-7fe7d1286a22\\" class=\\"\\">Videos:</p>
        <p id=\\"add510ff-2a30-45d6-837f-ac658c0a9f2e\\" class=\\"\\">
          â†’
          <a href=\\"https://www.coursera.org/learn/build-a-computer\\"
            >https://www.coursera.org/learn/build-a-computer</a
          >
        </p>
        <h3 id=\\"d515832a-2c38-4d27-9dfd-dda73cd8d7e1\\" class=\\"\\">Projects</h3>
        <ul id=\\"bafe2268-25fc-4afb-8752-eaad31af9f6c\\" class=\\"bulleted-list\\">
          <li>NAND to tetris computer</li>
        </ul>
        <ul id=\\"4b7ea969-52c8-4911-9512-e3af32d53848\\" class=\\"bulleted-list\\">
          <li>Chip 8 emulator</li>
        </ul>
        <figure id=\\"701fdbb8-602e-4a30-ba89-35a817c5b293\\" class=\\"link-to-page\\">
          <a href=\\"Computer%20Architecture/Game%20Boy%20Emulator.html\\"
            ><span class=\\"icon\\">ðŸ‘¾</span>Game Boy Emulator</a
          >
        </figure>
        <ul id=\\"68c18e60-e430-4971-8f72-e1730a23e1bd\\" class=\\"bulleted-list\\">
          <li>Contribute to virtual box etc</li>
        </ul>
        <ul id=\\"532467d4-63f4-4c39-9fb0-4b8623cb797a\\" class=\\"bulleted-list\\">
          <li>Write N64, PS1 emulator</li>
        </ul>
        <h3 id=\\"c99b6b2a-b5fa-4e30-b5d4-7bb1731d31af\\" class=\\"\\">Links</h3>
        <p id=\\"28e7bc1a-1722-415c-a56e-49911415d399\\" class=\\"\\">
          Ben eater:
          <a href=\\"https://www.youtube.com/channel/UCS0N5baNlQWJCUrhCEo8WlA\\"
            >https://www.youtube.com/channel/UCS0N5baNlQWJCUrhCEo8WlA</a
          >
        </p>
        <ul id=\\"e4d6b9ee-4adc-468c-a86c-b34ca8a9be14\\" class=\\"bulleted-list\\">
          <li>
            <a href=\\"https://www.youtube.com/watch?v=l7rce6IQDWs\\"
              >The world's worst video card?</a
            >
          </li>
        </ul>
        <p id=\\"637cb675-80b4-43a6-b769-3264b95f0bbc\\" class=\\"\\">
          <a href=\\"http://bitwisecmd.com/\\">http://bitwisecmd.com/</a>
        </p>
        <ul id=\\"d5980e65-5666-4924-b3e9-97bd44a614e0\\" class=\\"toggle\\">
          <li>
            <details open=\\"\\"
              ><summary>Websites related to electronics</summary>
              <p id=\\"c7d71dfe-0430-4dba-9eac-43df60b4d2ae\\" class=\\"\\">
                <a href=\\"https://www.hackster.io/\\">https://www.hackster.io/</a>
              </p>
              <p id=\\"92a4486f-0572-43a6-af0d-2fbd84367bbf\\" class=\\"\\">
                <a href=\\"https://www.instructables.com/\\"
                  >https://www.instructables.com/</a
                >
              </p>
              <p id=\\"5b06c805-4a97-43b3-9a89-7466fe8e086c\\" class=\\"\\">
                <a href=\\"https://www.adafruit.com/\\"
                  >https://www.adafruit.com/</a
                >
              </p></details
            >
          </li>
        </ul>
        <ul id=\\"180e82e0-9425-4fd9-8564-7203cc1168e8\\" class=\\"toggle\\">
          <li>
            <details open=\\"\\"
              ><summary>Console stuff</summary>
              <p id=\\"daae2625-5072-47b1-8ea2-0133ace1ba0d\\" class=\\"\\">
                <a
                  href=\\"https://internalregister.github.io/2019/03/14/Homebrew-Console.html\\"
                  >https://internalregister.github.io/2019/03/14/Homebrew-Console.html</a
                >
              </p>
              <p id=\\"9d1c43df-8d02-4dfc-919f-fa3b379cc682\\" class=\\"\\">
                <a href=\\"http://8bitworkshop.com/\\">http://8bitworkshop.com/</a>
              </p></details
            >
          </li>
        </ul>
        <ul id=\\"d07188b4-4261-447d-ab56-02ec13511bd2\\" class=\\"toggle\\">
          <li>
            <details open=\\"\\"
              ><summary>Breadboard stuff</summary>
              <ul
                id=\\"047913b9-046c-45fe-8e87-df24bdcbba80\\"
                class=\\"bulleted-list\\"
              >
                <li>
                  <a
                    href=\\"https://www.instructables.com/id/Zero-to-Breadboard-Simulation/\\"
                    >https://www.instructables.com/id/Zero-to-Breadboard-Simulation/</a
                  >
                </li>
              </ul>
              <ul
                id=\\"efbba93b-d4a4-4235-8bed-b0d81214252c\\"
                class=\\"bulleted-list\\"
              >
                <li>
                  <a href=\\"http://www.virtualbreadboard.com/html/index.html\\"
                    >http://www.virtualbreadboard.com/html/index.html</a
                  >
                </li>
              </ul>
              <ul
                id=\\"6c998cb4-7ed5-4ff6-b7ea-5890d4bc2a2a\\"
                class=\\"bulleted-list\\"
              >
                <li>
                  <a href=\\"https://www.tinkercad.com/circuits\\"
                    >https://www.tinkercad.com/circuits</a
                  >
                </li>
              </ul></details
            >
          </li>
        </ul>
        <ul id=\\"4ece6c8c-9588-44a9-8868-ac14e0e79413\\" class=\\"toggle\\">
          <li>
            <details open=\\"\\"
              ><summary>E-ink stuff</summary>
              <p id=\\"f34ba1fa-08b0-445f-b982-4420453f0bc6\\" class=\\"\\">
                <a
                  href=\\"https://www.raspberrypi.org/blog/using-e-ink-raspberry-pi/\\"
                  >https://www.raspberrypi.org/blog/using-e-ink-raspberry-pi/</a
                >
              </p>
              <p id=\\"407c7101-3b8c-4942-830c-09adaff6ad8d\\" class=\\"\\">
                <a
                  href=\\"https://hackaday.com/2017/04/20/e-ink-display-driven-diy/\\"
                  >https://hackaday.com/2017/04/20/e-ink-display-driven-diy/</a
                >
              </p></details
            >
          </li>
        </ul>
        <p id=\\"b50d5aa3-f4b4-43e9-ad75-50d35ef8bfb0\\" class=\\"\\">
          <a href=\\"https://www.nand2tetris.org/\\"
            >https://www.nand2tetris.org/</a
          >
        </p>
        <p id=\\"00a529cf-5394-463a-a558-5533dbbf4934\\" class=\\"\\"></p>
        <p id=\\"a2c8f510-66c1-46d9-8beb-f908085dff2f\\" class=\\"\\">
          <a href=\\"https://twitter.com/jessfraz/status/1199746662668754944\\"
            >https://twitter.com/jessfraz/status/1199746662668754944</a
          >
        </p>
        <p id=\\"5c36cd56-711f-4fb6-a29d-3f2b59f6e1e8\\" class=\\"\\">
          <a href=\\"https://www.youtube.com/channel/UCIwQ8uOeRFgOEvBLYc3kc3g\\"
            >https://www.youtube.com/channel/UCIwQ8uOeRFgOEvBLYc3kc3g</a
          >
        </p>
      </div>
    </article>
  </body>
</html>
"
`;
